<launch>
  <!-- Settings -->
  <arg name="gui" default="true" />
  <arg name="image_view" default="true" />
  <arg name="video_device" default="/dev/video0" />
  <arg name="threshold" default="235" doc="threshold value"/>
  <arg name="camera_type" default="usb" />
  <arg name="frame_rate" default="30.0" />

  <!-- Derived arguments to control which nodes to launch -->
  <arg name="launch_usb_cam" default="$(eval arg('camera_type') == 'usb')" />
  <arg name="launch_pylon_cam" default="$(eval arg('camera_type') == 'pylon')" />
  <arg name="image_topic_raw" default="/usb_cam/image_raw" if="$(eval arg('camera_type') == 'usb')" />
  <arg name="image_topic_raw" default="/pylon_camera_node/image_raw" unless="$(eval arg('camera_type') == 'usb')" />



  <!-- USB camera driver -->
  <node name="usb_cam" pkg="usb_cam" type="usb_cam_node" output="screen" if="$(arg launch_usb_cam)">
    <param name="video_device" value="$(arg video_device)" />
    <!--param name="image_width" value="1280" /-->
    <!--param name="image_height" value="720" /-->
    <param name="autoexposure" value="false" />
    <param name="pixel_format" value="yuyv" />
    <param name="camera_frame_id" value="usb_cam" />
    <param name="frame_rate" value="$(arg frame_rate)" />
    <param name="brightness" value="200" />
    <param name="autofocus" value="false" />
    <param name="io_method" value="mmap"/>
  </node>

  <!-- Pylon camera driver -->
  <include file="$(find pylon_camera)/launch/pylon_camera_node.launch" if="$(arg launch_pylon_cam)">
    <param name="frame_rate" value="$(arg frame_rate)" />
    <!-- <param name="gige/mtu_size" value="1500"/>
    <param name="grab_strategy" value="0"/>
    <param name="grab_timeout" value="500"/> -->
  </include>
 
  <!-- Meltpool Geometric Features -->
  <!-- <node name="coaxial_feature_extraction_node" pkg="coaxial_melt_pool_monitoring" type="coaxial_meltpool_feature_extraction.py">
    <param name="image"  value="$(arg image_topic_raw)" />
    <param name="threshold" value="$(arg threshold)"/>
  </node> -->

  <node name="coaxial_feature_extraction_node" pkg="coaxial_melt_pool_monitoring" type="coaxial_meltpool_feature_extraction_node">
    <param name="image"  value="$(arg image_topic_raw)" />
    <param name="threshold" value="$(arg threshold)"/>
  </node>

  <!-- image visualize -->
  <node name="image_view" pkg="image_view" type="image_view" respawn="false" output="screen" if="$(arg image_view)">
    <remap from="image" to="$(arg image_topic_raw)"/>
    <param name="autosize" value="true" />
  </node>

  <node name="image_view2" pkg="image_view" type="image_view" respawn="false" output="screen" if="$(arg image_view)">
    <remap from="image" to="/image_general_contour"/>
    <param name="autosize" value="true" />
  </node>

  <node name="image_view3" pkg="image_view" type="image_view" respawn="false" output="screen" if="$(arg image_view)">
    <remap from="image" to="/image_convex_hull"/>
    <param name="autosize" value="true" />
  </node>

  <node name="image_view4" pkg="image_view" type="image_view" respawn="false" output="screen" if="$(arg image_view)">
    <remap from="image" to="/image_moment_extract"/>
    <param name="autosize" value="true" />
  </node>

  

</launch>
